pkg/flowaggregator/flowaggregator.go

```diff
@@ -450,121 +273,42 @@ func (fa *flowAggregator) InitAggregationProcess() error {
 	return err
 }
 
-func (fa *flowAggregator) InitDBExportProcess(chInput clickhouseclient.ClickHouseInput) error {
-	var err error
-	fa.dbExportProcess, err = clickhouseclient.NewClickHouseClient(chInput)
-	return err
-}
-
-func (fa *flowAggregator) createAndSendTemplate(isRecordIPv6 bool) error {
-	templateID := fa.exportingProcess.NewTemplateID()
-	recordIPFamily := "IPv4"
-	if isRecordIPv6 {
-		recordIPFamily = "IPv6"
-	}
-	if isRecordIPv6 {
-		fa.templateIDv6 = templateID
-	} else {
-		fa.templateIDv4 = templateID
-	}
-	bytesSent, err := fa.sendTemplateSet(isRecordIPv6)
-	if err != nil {
-		fa.exportingProcess.CloseConnToCollector()
-		fa.exportingProcess = nil
-		fa.set.ResetSet()
-		return fmt.Errorf("sending %s template set failed, err: %v", recordIPFamily, err)
-	}
-	klog.V(2).InfoS("Exporting process initialized", "bytesSent", bytesSent, "templateSetIPFamily", recordIPFamily)
-	return nil
-}
-
-func (fa *flowAggregator) initExportingProcess() error {
-	// TODO: This code can be further simplified by changing the go-ipfix API to accept
-	// externalFlowCollectorAddr and externalFlowCollectorProto instead of net.Addr input.
-	var expInput exporter.ExporterInput
-	if fa.externalFlowCollectorProto == "tcp" {
-		// TCP transport does not need any tempRefTimeout, so sending 0.
-		expInput = exporter.ExporterInput{
-			CollectorAddress:    fa.externalFlowCollectorAddr,
-			CollectorProtocol:   fa.externalFlowCollectorProto,
-			ObservationDomainID: fa.observationDomainID,
-			TempRefTimeout:      0,
-			IsEncrypted:         false,
-			SendJSONRecord:      fa.sendJSONRecord,
-		}
-	} else {
-		// For UDP transport, hardcoding tempRefTimeout value as 1800s. So we will send out template every 30 minutes.
-		expInput = exporter.ExporterInput{
-			CollectorAddress:    fa.externalFlowCollectorAddr,
-			CollectorProtocol:   fa.externalFlowCollectorProto,
-			ObservationDomainID: fa.observationDomainID,
-			TempRefTimeout:      1800,
-			IsEncrypted:         false,
-			SendJSONRecord:      fa.sendJSONRecord,
-		}
-	}
-	ep, err := ipfix.NewIPFIXExportingProcess(expInput)
-	if err != nil {
-		return fmt.Errorf("got error when initializing IPFIX exporting process: %v", err)
-	}
-	fa.exportingProcess = ep
-	// Currently, we send two templates for IPv4 and IPv6 regardless of the IP families supported by cluster
-	if err = fa.createAndSendTemplate(false); err != nil {
-		return err
-	}
-	if err = fa.createAndSendTemplate(true); err != nil {
-		return err
-	}
-
-	return nil
-}
-
-func (fa *flowAggregator) Run(stopCh <-chan struct{}, wg *sync.WaitGroup) {
-	defer wg.Done()
+func (fa *flowAggregator) Run(stopCh <-chan struct{}) {
 	go fa.collectingProcess.Start()
 	defer fa.collectingProcess.Stop()
 	go fa.aggregationProcess.Start()
 	defer fa.aggregationProcess.Stop()
-	if fa.dbExportProcess != nil {
-		go fa.dbExportProcess.Start()
-		defer fa.dbExportProcess.Stop()
+	if fa.ipfixExporter != nil {
+		fa.ipfixExporter.Start()
+		defer fa.ipfixExporter.Stop()
```

Comment:
I think this line should be put inside the flowExportLoop (case <-stopCh). It will panic when we do enable clickhouse/ipfix after we had disabled clickhouse/ipfix. Because the fa.ipfixExporter and fa.clickHouseExporter won't be the original one.

Reply:
Thanks let me add some unit tests to confirm and I'll fix this.
I think the original code may have suffered from a similar issue. For example with the following sequence:
1) FA is started with ClickHouse enabled
2) CLI is used to disabled ClickHouse
3) FA is stopped (panic because `fa.dbExportProcess.Stop()` is called twice)

That being said, not a super serious bug since it only happens when stopping the FA

Reply:
@yuntanghsu Please check out the most recent commit. I made the change you suggested and also added a unit test. Let me know if there is a specific sequence of events you want me to test.

Reply:
It looks good now.